### PR TITLE
Add Gemm to ONNXToTOSA

### PR DESCRIPTION
**Pull request by flemairen6**

### CODE REVIEW

File: src/Conversion/ONNXToTOSA/Math/Gemm.cpp [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+        (adaptor.beta().convertToFloat() != 1.0F)) {
+      return false;
+    }
+    // Only Transpose B must be enabled
+    if (adaptor.transA() != 0 || adaptor.transB() != 1) {
+      return false;
+    }
+    return true;
+  }
+  LogicalResult matchAndRewrite(ONNXGemmOp op, OpAdaptor adaptor,
```
Comment (ljfitz) [inline src/Conversion/ONNXToTOSA/Math/Gemm.cpp:94]: I'd prefer this to be the 1st function defined in the class so it is easier for find.

File: src/Conversion/ONNXToTOSA/Math/Gemm.cpp [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+    Value A = adaptor.A();
+    Value B = adaptor.B();
+    Value C = adaptor.C();
+
+    Type resultType = getTypeConverter()->convertType(op.getResult().getType());
+
+    // If no bias is given to the GEMM operator, we create a 1D bias with all
+    // zeroes
+    if (C.getType().isa<mlir::NoneType>()) {
+      // B is supposed to have
```
Comment (ljfitz) [inline src/Conversion/ONNXToTOSA/Math/Gemm.cpp:46]: Missing some comment here?

File: src/Conversion/ONNXToTOSA/Math/Gemm.cpp [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+namespace onnx_mlir {
+
+namespace {
+
+
+
+class ONNXGemmOpLoweringToTOSA : public OpConversionPattern<ONNXGemmOp> {
+public:
+  using OpConversionPattern<ONNXGemmOp>::OpConversionPattern;
+  LogicalResult rewriteToTosaFC(ONNXGemmOp op, OpAdaptor adaptor,
```
Comment (ljfitz) [inline src/Conversion/ONNXToTOSA/Math/Gemm.cpp:35]: I wonder if splitting this in `check()`and `rewrite()`is helpful, or if a `matchAndRewriteToFC()` would work better.

File: src/Conversion/ONNXToTOSA/Math/Gemm.cpp [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+    }
+
+    rewriter.replaceOpWithNewOp<tosa::FullyConnectedOp>(op, resultType, A, B, C);
+    return success();
+  }
+
+  /// The GEMM can be described as a FullyConnected operator
+  /// Y = AB^T + C if we perform a transpose on B only with
+  /// alpha and beta factors set to 1.
+  /// The constraints on rank were taken from the torch implementation check
```
Comment (ljfitz) [inline src/Conversion/ONNXToTOSA/Math/Gemm.cpp:61]: This comment won't make much sense when upstreaming, as we don't plan to upstream our onnx-to-torch implementation.

File: src/Conversion/ONNXToTOSA/Math/Gemm.cpp [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+    if (!isCPresent) {
+      rewriter.eraseOp(C.getDefiningOp());
+    }
+
+    // If legal, create a FullyConnected operator instead
+    if (checkLegalTosaFCOp(op, adaptor)) {
+      return rewriteToTosaFC(op, adaptor, rewriter);
+    }
+    
+    // ONNX gives 2d matrix as input and expect a 2d output. TOSA expects everything to be 3D. As such, there
```
Comment (ljfitz) [inline src/Conversion/ONNXToTOSA/Math/Gemm.cpp:122]: I think the more common terminology here (pytorch etc.) is less 2D vs 3D, but rather that TOSA provides a batch matmul. Coming from ONNX the batch is set to one by definition, and not because it is unknown (it isn't unknown - we have a single batch to process).

File: src/Conversion/ONNXToTOSA/Math/Gemm.cpp [hunk 6]
```diff
@@ -0,0 +1,208 @@
...
+    // is a need to add reshapes operators before and after we do any computation. We set the batch as 1 as it
+    // is unknown.
+    llvm::SmallVector<int64_t> newShapeA{1};
+    llvm::SmallVector<int64_t> newShapeB{1};
+    newShapeA.append({shapeA[0], shapeA[1]});
+    newShapeB.append({shapeB[0], shapeB[1]});
```
Comment (ljfitz) [inline src/Conversion/ONNXToTOSA/Math/Gemm.cpp:128]:
```suggestion
    llvm::SmallVector<int64_t> newShapeA{1, shapeA[0], shapeA[1]};
    llvm::SmallVector<int64_t> newShapeB{1, shapeB[0], shapeB[1]};
```

File: src/Conversion/ONNXToTOSA/Math/Gemm.cpp [hunk 7]
```diff
@@ -0,0 +1,208 @@
...
+
+    A = tosa::CreateOpAndInfer<tosa::ReshapeOp>(
+        rewriter, op->getLoc(),
+        UnrankedTensorType::get(AType.getElementType()), A,
```
Comment (ljfitz) [inline src/Conversion/ONNXToTOSA/Math/Gemm.cpp:132]: While `UnrankedTensorType` is allowed, I read https://github.com/llvm/llvm-project/blob/main/mlir/include/mlir/Dialect/Tosa/IR/TosaTypesBase.td#L123 as suggesting it is preferable to use a ranked tensor if you can do so. I'm not an expert here, so if you can point out why unranked would be preferred I'd be happy to learn.

File: src/Conversion/ONNXToTOSA/Math/Gemm.cpp [hunk 8]
```diff
@@ -0,0 +1,208 @@
...
+      B = tosa::CreateOpAndInfer<tosa::TransposeOp>(rewriter, loc, outputType, B, targetTensor).getResult();
+    }
+    
+    Value alphaMulResult = NULL;
+    Value betaMulResult = NULL;
+    // If Alpha is present and not 1, we create a multiply operation for alpha * A
+    if (alpha && alpha.getValueAsDouble() != 1.) {
+      alphaMulResult = tosa::CreateOpAndInfer<tosa::MulOp>(rewriter, loc,
+            tosaResult,
+            tosa::getTosaConstTensorSingleF32(rewriter, op, alpha.getValueAsDouble(), newShapeA),
```
Comment (ljfitz) [inline src/Conversion/ONNXToTOSA/Math/Gemm.cpp:159]: @maxbartel - was it in places like this that you had trouble with `double` and `F32`?

File: src/Conversion/ONNXToTOSA/Math/Gemm.cpp [hunk 9]
```diff
@@ -0,0 +1,208 @@
...
+      A = tosa::CreateOpAndInfer<tosa::TransposeOp>(rewriter, loc, outputType, A, targetTensor).getResult();
+    }
+    if (transB) {
+      Value targetTensor = mlir::tosa::getConstTensor<int32_t>(rewriter, op, {0, 2, 1}, {3}).value();
+      Type outputType = UnrankedTensorType::get(AType.getElementType());
+      B = tosa::CreateOpAndInfer<tosa::TransposeOp>(rewriter, loc, outputType, B, targetTensor).getResult();
+    }
+    
+    Value alphaMulResult = NULL;
+    Value betaMulResult = NULL;
```
Comment (ljfitz) [inline src/Conversion/ONNXToTOSA/Math/Gemm.cpp:154]: If you initialized these to `A` and `C` respectively, the logic further down would be simplified.

File: src/Conversion/ONNXToTOSA/Math/Gemm.cpp [hunk 10]
```diff
@@ -0,0 +1,208 @@
...
+      addRes = tosa::CreateOpAndInfer<tosa::AddOp>(rewriter, loc, tosaResult, matmulRes, betaMulResult).getResult();
+    }
+    else if (isCPresent) {
+      addRes = tosa::CreateOpAndInfer<tosa::AddOp>(rewriter, loc, tosaResult, matmulRes, adaptor.C()).getResult();
+    }
+    else {
+      addRes = matmulRes;
+    }
+
+    // Add reshape to go back to the original size
```
Comment (ljfitz) [inline src/Conversion/ONNXToTOSA/Math/Gemm.cpp:192]:
```suggestion
    // Add reshape to go back to the original shape.
```

File: src/Conversion/ONNXToTOSA/ONNXToTOSALegalizeUtils.cpp [hunk 11]
```diff
@@ -45,5 +45,65 @@ Value getTosaConstTensorSingleF32(PatternRewriter &rewriter, Operation *op,
   auto constOp = rewriter.create<ConstOp>(op->getLoc(), constType, constAttr);
   return constOp.getResult();
 }
+
+// Templated function to create a constant op for given type and shape.
+// T: storage C type.
+// Default template creates a constant tensor in T.
+template <typename T>
+llvm::Optional<Value> getConstTensor(PatternRewriter &rewriter, Operation *op,
```
Comment (ljfitz) [inline src/Conversion/ONNXToTOSA/ONNXToTOSALegalizeUtils.cpp:53]: `Value` already had nullptr-like semantics - the `Optional` is overkill. But I guess this comes from upstream?

File: src/Conversion/ONNXToTOSA/Tensor/Constant.cpp [hunk 12]
```diff
@@ -16,6 +16,7 @@
 #include "mlir/IR/BuiltinAttributeInterfaces.h"
 #include "mlir/IR/TypeUtilities.h"
 #include "src/Conversion/ONNXToTOSA/ONNXToTOSACommon.hpp"
+#include <cassert>
```
Comment (ljfitz) [inline src/Conversion/ONNXToTOSA/Tensor/Constant.cpp:19]: Why was this added?

File: test/mlir/conversion/onnx_to_tosa/Math/Gemm_to_linear.mlir [hunk 13]
```diff
@@ -0,0 +1,23 @@
...
+// create a bias with values of 0. Our testing relies on the lowering from torch to host code and this
+// will allow us to run the lowering passes.
+module attributes {}  {
+  func.func @Gemm_to_linear(%arg0: tensor<1x5xf32>, %arg1: tensor<4x5xf32>, %arg2: tensor<4xf32>) -> tensor<1x4xf32> attributes {input_names = ["a", "b"], output_names = ["y"]} {
+    %0 = "onnx.Gemm"(%arg0, %arg1, %arg2) {transB = 1 : si64} : (tensor<1x5xf32>, tensor<4x5xf32>, tensor<4xf32>) -> tensor<1x4xf32>
+//CHECK-LABEL:  @Gemm_to_linear(%arg0: tensor<1x5xf32>, %arg1: tensor<4x5xf32>, %arg2: tensor<4xf32>) -> tensor<1x4xf32>
+//CHECK-DAG:    "tosa.fully_connected"(%arg0, %arg1, %arg2) : (tensor<1x5xf32>, tensor<4x5xf32>, tensor<4xf32>) -> tensor<1x4xf32>
+    return %0 : tensor<1x4xf32>
+  }
+
```
Comment (ljfitz) [inline test/mlir/conversion/onnx_to_tosa/Math/Gemm_to_linear.mlir:13]: Please add `// -----` markers for the `-split-input-file` to work.

File: test/mlir/conversion/onnx_to_tosa/Math/Gemm_to_matmul_allattr.mlir [hunk 14]
```diff
@@ -0,0 +1,21 @@
+// RUN: onnx-mlir-opt --shape-inference --convert-onnx-to-tosa %s -split-input-file | FileCheck %s
```
Comment (ljfitz) [inline test/mlir/conversion/onnx_to_tosa/Math/Gemm_to_matmul_allattr.mlir:1]: All `gemm_to_matmul` related tests can go int he one file, given the `-split-input-file` support.

**Review by ljfitz — COMMENTED** [review]
A lot of good work in here @flemairen6. There are some issues to resolve, such as care for dynamic shapes and whether using `UnrankedTensorType` is really preferable. Other than that the comments are largely in the area of style.

I didn't get to give the tests a good looking over yet. @maxbartel - I hope you can pay special attention to these when reviewing.

File: src/Conversion/ONNXToTOSA/ONNXToTOSALegalizeUtils.cpp (hunk 11 above)
Comment (flemairen6) [reply src/Conversion/ONNXToTOSA/ONNXToTOSALegalizeUtils.cpp:53]: Yes, this comes from upstream

File: src/Conversion/ONNXToTOSA/Math/Gemm.cpp (hunk 8 above)
Comment (maxbartel) [reply src/Conversion/ONNXToTOSA/Math/Gemm.cpp:159]: Yes exactly. This line works because we moved the double to float cast to the function arguments.

File: src/Conversion/ONNXToTOSA/Math/Gemm.cpp (hunk 7 above)
Comment (flemairen6) [reply src/Conversion/ONNXToTOSA/Math/Gemm.cpp:74]: I just think it's easier because I don't have to compute the shape myself, and CreateOpAndInfer already runs shape inference to resize the tensors as wanted

Comment (maxbartel) [reply src/Conversion/ONNXToTOSA/Math/Gemm.cpp:74]: `RankedTensorType::get({-1, -1, -1}, AType.getElementType())` is an alternative, but you at least check that the tensor has the right rank.

File: src/Conversion/ONNXToTOSA/ConvertONNXToTOSA.cpp [hunk 15]
```diff
@@ -14,6 +14,7 @@
 
 #include "mlir/IR/BuiltinTypes.h"
 #include "src/Conversion/ONNXToTOSA/ONNXToTOSACommon.hpp"
+#include "mlir/Transforms/DialectConversion.h"
```
Comment (maxbartel) [inline src/Conversion/ONNXToTOSA/ConvertONNXToTOSA.cpp:17]: I don't think you need this header

File: test/mlir/conversion/onnx_to_tosa/Math/Gemm_to_linear.mlir [hunk 16]
```diff
@@ -0,0 +1,22 @@
+// RUN: onnx-mlir-opt --shape-inference --convert-onnx-to-tosa %s -split-input-file | FileCheck %s
+
+// because the optional bias is not yet implemented in the upstream TorchToLinAlg pass we will instead
+// create a bias with values of 0. Our testing relies on the lowering from torch to host code and this
```
Comment (maxbartel) [inline test/mlir/conversion/onnx_to_tosa/Math/Gemm_to_linear.mlir:4]: It think these lines are wrongly copied from the onnx to torch test.

File: test/mlir/conversion/onnx_to_tosa/Math/Gemm_to_matmul.mlir [hunk 17]
```diff
@@ -0,0 +1,117 @@
...
+// CHECK-DAG:  %[[A:.*]] = "tosa.reshape"(%arg0) {new_shape = [1, 3, 5]} : (tensor<3x5xf32>) -> tensor<1x3x5xf32>
+// CHECK-DAG:  %[[B:.*]] = "tosa.reshape"(%arg1) {new_shape = [1, 5, 4]} : (tensor<5x4xf32>) -> tensor<1x5x4xf32>
+// CHECK-DAG:  %[[AB:.*]] = "tosa.matmul"(%[[A]], %[[B]]) : (tensor<1x3x5xf32>, tensor<1x5x4xf32>) -> tensor<1x3x4xf32>
+// CHECK-DAG:  %[[ABC:.*]] = "tosa.add"(%[[AB]], %arg2) : (tensor<1x3x4xf32>, tensor<3x4xf32>) -> tensor<1x3x4xf32>
+// CHECK-DAG:  %[[RES:.*]] = "tosa.reshape"(%[[ABC]]) {new_shape = [3, 4]} : (tensor<1x3x4xf32>) -> tensor<3x4xf32>
+// CHECK-DAG:  return %[[RES]] : tensor<3x4xf32>
+
+// -----
+
+func.func @test_alpha(%arg0: tensor<3x6xf32>, %arg1: tensor<6x4xf32>, %arg2: tensor<3x4xf32>) -> tensor<3x4xf32> attributes {input_names = ["a", "b", "c"], output_names = ["y"]} {
```
Comment (maxbartel) [inline test/mlir/conversion/onnx_to_tosa/Math/Gemm_to_matmul.mlir:17]: I don't think the attributes at the end add to the test. Maybe better delete them?

File: test/mlir/conversion/onnx_to_tosa/Math/Gemm_to_matmul.mlir [hunk 18]
```diff
@@ -0,0 +1,117 @@
...
+  
+func.func @test_transB(%arg0: tensor<3x6xf32>, %arg1: tensor<4x6xf32>, %arg2: tensor<3x4xf32>) -> tensor<3x4xf32> attributes {input_names = ["a", "b", "c"], output_names = ["y"]} {
+  %0 = "onnx.Gemm"(%arg0, %arg1, %arg2) {alpha = 1.184 : f32, transB = 1 : si64} : (tensor<3x6xf32>, tensor<4x6xf32>, tensor<3x4xf32>) -> tensor<3x4xf32>
+  return %0 : tensor<3x4xf32>
+}
+// CHECK-LABEL:  @test_transB(%arg0: tensor<3x6xf32>, %arg1: tensor<4x6xf32>, %arg2: tensor<3x4xf32>) -> tensor<3x4xf32>
+// CHECK-DAG:    %[[A:.*]] = "tosa.reshape"(%arg0) {new_shape = [1, 3, 6]} : (tensor<3x6xf32>) -> tensor<1x3x6xf32>
+// CHECK-DAG:    %[[B:.*]] = "tosa.reshape"(%arg1) {new_shape = [1, 4, 6]} : (tensor<4x6xf32>) -> tensor<1x4x6xf32>
+// CHECK-DAG:    "tosa.const"() {value = dense<[0, 2, 1]> : tensor<3xi32>} : () -> tensor<3xi32>
+// CHECK-DAG:    %[[TRANB:.*]] = "tosa.transpose"(%[[B]], %2) : (tensor<1x4x6xf32>, tensor<3xi32>) -> tensor<1x6x4xf32>
```
Comment (maxbartel) [inline test/mlir/conversion/onnx_to_tosa/Math/Gemm_to_matmul.mlir:73]: Same as above

File: test/mlir/conversion/onnx_to_tosa/Math/Gemm_to_matmul.mlir [hunk 19]
```diff
@@ -0,0 +1,117 @@
...
+
+func.func @test_transA(%arg0: tensor<6x3xf32>, %arg1: tensor<6x4xf32>, %arg2: tensor<3x4xf32>) -> tensor<3x4xf32> attributes {input_names = ["a", "b", "c"], output_names = ["y"]} {
+  %0 = "onnx.Gemm"(%arg0, %arg1, %arg2) {transA = 1 : si64} : (tensor<6x3xf32>, tensor<6x4xf32>, tensor<3x4xf32>) -> tensor<3x4xf32>
+  return %0 : tensor<3x4xf32>
+}
+// CHECK-LABEL:  @test_transA(%arg0: tensor<6x3xf32>, %arg1: tensor<6x4xf32>, %arg2: tensor<3x4xf32>) -> tensor<3x4xf32> attributes {input_names = ["a", "b", "c"], output_names = ["y"]} {
+// CHECK-DAG:    %[[A:.*]] = "tosa.reshape"(%arg0) {new_shape = [1, 6, 3]} : (tensor<6x3xf32>) -> tensor<1x6x3xf32>
+// CHECK-DAG:    %[[B:.*]] = "tosa.reshape"(%arg1) {new_shape = [1, 6, 4]} : (tensor<6x4xf32>) -> tensor<1x6x4xf32>
+// CHECK-DAG:    "tosa.const"() {value = dense<[0, 2, 1]> : tensor<3xi32>} : () -> tensor<3xi32>
+// CHECK-DAG:    %[[TRANA:.*]] = "tosa.transpose"(%[[A]], %2) : (tensor<1x6x3xf32>, tensor<3xi32>) -> tensor<1x3x6xf32>
```
[the remark this answers is not in the document]
Comment (flemairen6) [reply test/mlir/conversion/onnx_to_tosa/Math/Gemm_to_matmul.mlir:57]: It's just a constant that's used on the next line, I don't think it really matters.

File: src/Conversion/ONNXToTOSA/Math/Gemm.cpp (hunk 7 above)
Comment (flemairen6) [reply src/Conversion/ONNXToTOSA/Math/Gemm.cpp:74]: Both are fine by me, if you think it's better I don't have a strong opinion

**Review by maxbartel — APPROVED** [review]
